multitenant: exclude background job reads/writes from tenant cost con…

andy-kimball opened the pull request:
Previously, reads and writes issued to the Storage layer by the job system
and by background jobs were included in tenant accounting. This is not ideal,
since we make a best effort to exclude operations from costing if they are
not under user control. This commit excludes the activities of the job system
from cost control, as well as the following jobs:

  - Long-running migration job
  - SQL stats compaction
  - Span config reconciler

None of these jobs are triggered by user actions and none can be disabled by
users.

NOTE: A cost control exemption does not exclude CPU or Egress costs from
accounting, since those cannot be attributed to individual jobs.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83005)
<!-- Reviewable:end -->

miretskiy commented on pkg/jobs/job_scheduler.go:
```diff
@@ -547,6 +548,11 @@ func StartJobSchedulerDaemon(
+	// Since the job scheduler system is not under user control, exclude it from
+	// from cost accounting and control. Individual jobs are not part of this
+	// exclusion.
+	ctx = multitenant.WithTenantCostControlExemption(ctx)
```
If I understand this mechanism correctly, any action performed with context with cost control exemption marker will be excluded from cost accounting.

It looks like jobs system ensures that's the case by verify if the job (resumer) marked itself as exempt.  However, the way scheduled jobs are written ATM, there is nothing that requires a schedule implementation to spawn a job.  What the schedule does -- is up to the schedule implementation. At this time -- all schedule implementations do spawn jobs.  But I wonder if we should be defensive and clear out exemption marker right before we execute the schedule (https://github.com/cockroachdb/cockroach/blob/master/pkg/jobs/job_scheduler.go#L184-L184).

andy-kimball replied:
I did have it that way originally, but it was more complex because I needed to add a "don't exempt" value to the context. I want to be able to back-port this, and it was getting rather complex, so I backed off and stuck with what you see. I'm not super-worried about job schedulers that don't schedule jobs - we just need to make sure we're charging for the 99% cases, and not get too hung up on the 1% case.

miretskiy commented on pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go:
You might find TakeOverJobScheduling more convenient -- just returns you a function you can invoke directly, without depending on any pacing logic in the scheduler.

andy-kimball replied:
OK, I switched to that.

miretskiy commented on pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go:
yeah -- I think having direct function you can invoke might make this logic a bit easier...  You might also consider just invoking this function many times -- to make sure that nothing is reported by scheduled jobs daemon itself.
But it's up to you.

andy-kimball replied:
Done.

miretskiy reviewed (approved):
mostly nits, and just 1 question.

andy-kimball committed:
multitenant: exclude background job reads/writes from tenant cost control

Previously, reads and writes issued to the Storage layer by the job system
and by background jobs were included in tenant accounting. This is not ideal,
since we make a best effort to exclude operations from costing if they are
not under user control. This commit excludes the activities of the job system
from cost control, as well as the following jobs:

  - Long-running migration job
  - SQL stats compaction
  - Span config reconciler

None of these jobs are triggered by user actions and none can be disabled by
users.

NOTE: A cost control exemption does not exclude CPU or Egress costs from
accounting, since those cannot be attributed to individual jobs.

Release note: None
